[1,2단계 - 로또] 김민지 미션 제출합니다

mmm307955 opened the pull request:
안녕하세요! 이번부터는 많이 어려워졌다고 느꼈는데요... 😢 
배울 게 많아서 그렇다고 생각합니다! 이번 리뷰도 잘 부탁드려요!

## 구현 내용
- LottoApplication을 시작점으로 사용하여 로또 자동 생성 기능 구현
- 사용자로부터 구입 금액과 지난 주 당첨 번호를 입력받고, 수익률을 계산하여 출력하는 흐름 구성
- 로또 티켓, 번호, 금액, 수익률 계산 로직을 도메인 객체로 분리
- LottoNumbers: List<Integer>를 감싸 유효성 검증 및 번호 일치 계산 책임 부여
- LottoTickets: List<LottoTicket>을 일급 컬렉션으로 래핑하여 각 티켓의 일치 개수를 집계하는 책임 수행

## 리팩토링 및 고민한 부분
- main 클래스명을 LottoController → LottoApplication으로 변경하는 것이 더 적절할 것 같아 수정했습니다!
- LottoNumbers 생성 시 중복 값, 개수, 범위(1~45) 유효성 검증을 추가했습니다.

## 질문사항
- 수익률 결과가 0.35처럼 소수점 이하 두 자리로 출력되는데, 단순히 소수점 아래를 자른 걸까요?
  (현재는 `%.2f` 포맷팅으로 출력하고 있습니다!)

## ETC
일급 컬렉션에 대해서 예진님이 질문을 주시고, 글렌이 추가로 설명을 해 주셔서 더 명확히 이해할 수 있었어요!
이전에는 단순히 Collection을 Wrapping하고 다른 멤버변수가 없는 객체, 혹은 가독성을 높이기 위한 수단(int money → Money money) 정도로만 생각했어요.
두 분의 대화를 통해서 (검증 로직과 책임이 무조건 있어야 하는 건 아니지만), 그 안에서 스스로 검증 로직을 가질 수 있는, 책임 있는 하나의 객체라는 걸 알게 되었습니다. 이제는 일급 컬렉션이 “단단하게 완성된 하나의 부품” 같은 느낌으로 다가옵니다!

mmm307955 committed:
feat: 1단계 - 로또 자동 구매 기능 구현

mmm307955 committed:
feat: 2단계 - 당첨 결과 및 수익률 계산 기능 구현

mmm307955 committed:
test: Lotto 도메인 클래스 단위 테스트 추가

mmm307955 committed:
docs: README 내용 수정

mmm307955 committed:
refactor: main 메서드를 LottoApplication으로 이동 및 테스트 코드 추가

seokjin8678 commented on src/main/java/LottoApplication.java:
```diff
@@ -0,0 +1,34 @@
+
+public class LottoApplication {
+    public static void main(String[] args) {
+        Money money = new Money(InputView.inputMoney());
+        LottoMachine lottoMachine = new LottoMachine();
+        int ticketCount = money.getTicketCount();
+        List<LottoTicket> tickets = lottoMachine.generateTickets(ticketCount);
+
+        OutputView.printTicketCount(ticketCount);
+        OutputView.printTickets(tickets);
+
+        List<Integer> winningNumbersInput = InputView.inputWinningNumbers();
+        LottoNumbers winningNumbers = new LottoNumbers(winningNumbersInput);
+
+        LottoTickets lottoTickets = new LottoTickets(tickets);
+        Map<Integer, Integer> matchResults = lottoTickets.countMatchResults(winningNumbers);
+
+        OutputView.printResult(matchResults);
+
+        int totalPrize = LottoResult.calculateTotalPrize(matchResults);
+        double profitRate = LottoResult.calculateRateOfReturn(totalPrize, money.getAmount());
+        OutputView.printProfitRate(profitRate);
```
`함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.`

mmm307955 replied:
`LottoController` 클래스 추가하여 메서드 길이 맞추었습니다!

seokjin8678 commented on src/test/java/domain/LottoMachineTest.java:
```diff
@@ -0,0 +1,36 @@
+        assertThat(tickets).hasSize(count);
+    }
+
+    @Test
+    @DisplayName("생성된 로또 티켓은 6개의 숫자를 가진다")
+    void eachTicket_has_SixNumbers() {
+        //Given
+        LottoMachine lottoMachine = new LottoMachine();
+        List<LottoTicket> lottoTickets = lottoMachine.generateTickets(10);
+
+        for (LottoTicket lottoTicket : lottoTickets) {
+            List<Integer> numbers = lottoTicket.getNumbers();
+            assertThat(numbers).hasSize(6);
+        }
+    }
```
검증을 `10번` 수행하는 이유가 있을까요?

mmm307955 replied:
로또 티켓을 생성하는 과정에 랜덤한 요소를 포함하고 있기 때문에 단일 케이스만 테스트하면 우연히 통과할 가능성이 있을 것 같아서, 여러 개(10개)의 로또 티켓을 생성해 각 티켓에 6개의 숫자를 가지고 있는지 테스트하려 했습니다!
물론 이 테스트가 로직 전반을 완전히 커버하진 않지만, 단순히 1개만 테스트하는 것보다 의미 있는 검증이라고 생각했어요.

seokjin8678 replied:
해당 테스트를 실행하는 시점에 운이 너무 좋아서, 10개 모두 우연히 통과하면 어떻게 하실건가요?

그렇다면 시도 횟수를 100번으로 늘려야 할까요?

테스트 할 수 없는 코드를 억지로 테스트 할 필요가 있을까요?

mmm307955 replied:
그러네요... 공감합니다. 테스트할 수 없는 코드를 억지로 검증하려다보면 오히려 신뢰도가 떨어지는 것 같다고 생각하여 해당 테스트는 제거하였습니다!

seokjin8678 commented on src/test/java/domain/LottoNumbersTest.java:
```diff
@@ -0,0 +1,73 @@
+            .hasMessage("로또 번호는 6개여야 합니다.");
+    }
+
+    @Test
+    @DisplayName("로또 번호에 1보다 작거나 45보다 큰 숫자가 있으면 예외가 발생한다")
+    void throwsException_when_input_LottoNum_outOfRange() {
+        //Given
+        List<Integer> numbers = List.of(0, 2, 3, 4, 5, 46);
+
+        //When & Then
+        assertThatThrownBy(() -> new LottoNumbers(numbers))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessage("로또 번호는 1~45 사이여야 합니다.");
+    }
```
다음과 같이 `@ParameterizedTest`를 사용하면, 더 명확할 것 같네요!
```suggestion
    @ValueSource(ints = {0, 46})
    @ParameterizedTest
    @DisplayName("로또 번호에 1보다 작거나 45보다 큰 숫자가 있으면 예외가 발생한다")
    void throwsException_when_input_LottoNum_outOfRange(int illegalNumber) {
        //Given
        List<Integer> numbers = List.of(1, 2, 3, 4, 5, illegalNumber);

        //When & Then
        assertThatThrownBy(() -> new LottoNumbers(numbers))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("로또 번호는 1~45 사이여야 합니다.");
    }
```

mmm307955 replied:
오오...! 😲 `illegalNumber`로 표현해서 정말 더 명확하네요!! 이렇게 바로 적용했습니다!

seokjin8678 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,32 @@
+    );
+
+    public static int calculateTotalPrize(Map<Integer, Integer> matchResults) {
+        int totalPrize = 0;
```
1등이 2번 당첨되면 어떻게 되나요?

mmm307955 replied:
```
3개 일치 (5000원) - 44803개
4개 일치 (50000원) - 2758개
5개 일치 (1500000원) - 54개
6개 일치 (2000000000원) - 1개
총 수익률은 -0.93입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)
```
헉... 그러고 보니 int 범위를 초과해서 -로 돌아가겠네요... long 자료형으로 수정했습니다!
1등 2번 당첨은 아니지만... 콘솔에서는 이렇게 나오네요! 😂

seokjin8678 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,32 @@
+        6, 2000000000
+    );
+
+    public static int calculateTotalPrize(Map<Integer, Integer> matchResults) {
```
Map의 `Key`, `Value`가 무엇인지 알 수 있나요?

mmm307955 replied:
```
//로또 당첨 결과: Rank별 당첨 개수
    public static Prize calculateTotalPrize(Map<Rank,Integer> matchResults) {
```
`Rank` enum을 만들고 이렇게 수정해보았습니다!

seokjin8678 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,32 @@
+        throw new AssertionError("LottoResult는 인스턴스화 할 수 없습니다.");
+    }
+
+    private static final Map<Integer, Integer> PRIZE_TABLE = Map.of(
+        3, 5000,
+        4, 50000,
+        5, 1500000,
+        6, 2000000000
+    );
```
마찬가지로, Map의 Key, Value가 무엇인지 알 수 있나요?

mmm307955 replied:
처음에는 `Map<Integer, Integer>` 대신 `<MatchCount, Prize>`처럼 도메인 객체로 감싸는 방법을 생각해보았고, 이후에는 `enum`을 사용하는 방식도 함께 검토해보았어요.

고민해본 결과, `enum`을 사용하면 컴파일 시점에 허용된 값만 사용하도록 강제할 수 있어 타입 안정성이 높아지고,
상금 계산과 같은 비즈니스 로직도 함께 캡슐화할 수 있어서 코드가 더 간결하고 유지보수하기 쉬워질 것 같다고 판단했습니다.

따라서 `enum`을 활용한 방향으로 리팩토링해보려고 합니다!

seokjin8678 commented on src/main/java/domain/Money.java:
```diff
@@ -0,0 +1,21 @@
+package domain;
+
+public class Money {
+    private  static final int TICKET_PRICE = 1000;
```
```suggestion
    private static final int TICKET_PRICE = 1000;
```

mmm307955 replied:
어멋! 수정했습니다

seokjin8678 commented on src/main/java/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,63 @@
+        return numbers.size() != new HashSet<>(numbers).size();
+    }
+
+    public List<Integer> getNumbers() {
+        return List.copyOf(numbers);
+    }
```
방어적 복사를 수행하셨군요!

`Collections.unmodifiableList()` 메서드를 사용하는 방법도 있으셨을 텐데, 왜 `List.copyOf`를 사용하셨나요?

mmm307955 replied:
사실 처음에는 두 메서드의 차이를 명확히 알지 못하고 사용했습니다. 😅
리뷰 덕분에 이번 기회에 차이를 공부해보았는데요,

`Collections.unmodifiableList(list)`는 리스트 수정을 막아주긴 하지만, 원본 리스트를 그대로 참조합니다. 그래서 외부에서 `list.set()` 같은 방식으로 원본을 바꾸면, 이 리스트도 함께 변경되는 위험이 있습니다.

반면 `List.copyOf(list)`는 리스트를 복사해서 새로운 불변 리스트를 만들기 때문에, 원본과 완전히 분리되어 안전하다고 생각했습니다. 또한 `null` 값도 허용하지 않아서 더 엄격한 불변성을 보장해줍니다.

이번에는 외부에서 리스트를 전달받아도 내부 상태가 변하지 않도록 방어적 복사를 위해 `List.copyOf()`를 선택했습니다! 🙇‍♀️

seokjin8678 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,47 @@
+        System.out.printf("총 수익률은 %.2f입니다.%s%n", profitRate, statusMessage);
+    }
+
+    private static String getProfitStatusMessage(double profitRate) {
+        if (profitRate < 1.0) {
+            return "(기준이 1이기 때문에 결과적으로 손해라는 의미임)";
+        }
+        return "(기준이 1이기 때문에 결과적으로 이익이라는 의미임)";
+    }
```
이 출력문 까지 구현해주셨군요.. 😂😂😂

seokjin8678 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,24 @@
+
+    public int countMatch(LottoNumbers winningNumbers) {
+        return numbers.countMatch(winningNumbers);
+    }
```
int의 범위는 몇인가요?

그렇다면 당첨될 수 있는 개수의 범위도 int의 범위를 따라가지 않을까요?

mmm307955 replied:
이 메서드는 하나의 로또 티켓이 당첨 번호와 몇 개 일치했는지를 반환하는 메서드이기 때문에 일치 개수는 로또 번호 수(6개(0 ~ 6 사이의 정수))를 넘을 수 없어 반환 타입이 `int`면 적절하다고 생각했습니다! 제가 이해한 게 맞을까요?

seokjin8678 replied:
이 클래스를 사용하는 곳에서도 같은 생각을 할 수 있을까요?

`등수` 같은 객체를 반환했으면 더 명확하게 표현될 수 있지 않았을까요?

mmm307955 replied:
의미 전달을 위해 `Rank` enum을 반환하도록 수정했습니다! 
일치 개수를 직접 사용하는 것보다 등수 객체를 만환하는 방식이 의도를 더 명확하게 표현할 수 있는 것 같습니다!

seokjin8678 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public String toString() {
+        return numbers.toString();
+    }
```
왜 toString을 재정의 해주셨나요?

mmm307955 replied:
`toString()`을 재정의하지 않으면 `LottoTicket` 객체를 출력할 때 다음과 같이 클래스명 + 해시코드로 이루어진 문자열이 출력됩니다. 어떤 번호가 포함됐는지 로또 번호 목록을 콘솔 출력에서도 확인할 수 있게 하기 위해 `toString()`을 재정의했습니다!

```
3개를 구매했습니다.
domain.LottoTicket@3af49f1c
domain.LottoTicket@19469ea2
domain.LottoTicket@13221655
```

seokjin8678 replied:
콘솔 출력은 어느 계층에서 수행하는 일인가요?

그렇다면 출력 형식이 `[1, 2, 3]`이 아닌 `(1, 2, 3)`으로 변경되어야 할 때, 어떤 계층을 수정해야 하나요?

또한, 어떤 개발자가 재구현한 `toString`을 다른 형식으로 변경하거나, 중요하지 않아 보여 삭제한다면 어떤 일이 발생할 것 같나요?

mmm307955 replied:
그렇네요! 처음에는 해당 클래스 내부에서 `toString()`을 재정의해서 출력 형식을 지정하는 것이 자연스럽다고 생각했지만,
출력의 책임은 View 계층인 `OutputView`가 가지는 것이 더 적절하다는 점을 알게 되었습니다!

만약 `LottoTicket` 클래스 내부의 `toString()`에 출력 형식을 의존하게 되면 다른 개발자가 해당 메서드를 변경하거나 삭제했을 때 콘솔 출력이 깨지는 등 예기치 못한 문제가 발생할 수 있을 것 같아요.

seokjin8678 commented on src/main/java/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,63 @@
+            .count();
+    }
+
+    private void validate(List<Integer> numbers) {
+        validateSize(numbers);
+        validateRange(numbers);
+        validateDuplicateNumber(numbers);
+    }
```
꼼꼼한 검증 로직 👍👍

seokjin8678 commented on src/main/java/domain/Money.java:
```diff
@@ -0,0 +1,21 @@
+package domain;
+
+public class Money {
```
`int money = 1000` 대신 `Money money = new Money(1000)`과 같이 사용하면, 명확하게 돈이라는 의미를 가지게 할 수 있겠군요!

또한 내부 필드가 final으로 지정되어, 여러 스레드에서 공유해도 안전할 것 같네요!

이러한 객체 구현을 "값 객체(Value Object)"라고 불러요.

값 객체는 상태를 변경 메서드를 제공하지 않으며, 그 자체로 식별성을 가져요.

하지만 지금 구현에서도 그럴까요?

```java
Money first = new Money(100);
Money second = new Money(100);

assertThat(first).isEqualTo(second);
```

mmm307955 replied:
현재 구현에서는 `equals()`와 `hashCode()`를 재정의하지 않아,
값이 같더라도 서로 다른 객체로 판단되고 있습니다.

값 객체로서의 역할을 명확히 하기 위해 `equals()`와 `hashCode()`를 재정의하여
값 기반 비교가 가능하도록 수정하겠습니다!

seokjin8678 commented on src/main/java/domain/Money.java:
```diff
@@ -0,0 +1,21 @@
+    private  static final int TICKET_PRICE = 1000;
+    private final int amount;
+
+    public Money(int amount) {
+        if (amount < TICKET_PRICE) {
+            throw new IllegalArgumentException("금액은 1000원 이상이어야 합니다.");
+        }
+        this.amount = amount;
+    }
```
왜 Money는 1000원 이상이면 안 되나요?

mmm307955 replied:
로또 한 장 가격이 1000원이기 때문에, 1000원 미만의 금액에 대해서는 예외를 던지도록 했습니다...! 이걸 말씀하시는 게 맞을까요?

seokjin8678 replied:
`Money`라는 클래스는 로또를 구매하는 것 외에도 당첨금에도 사용될 수 있을 것 같아요.

그렇다면 이러한 검증 로직이 있어야 할 곳은 `Money`가 아닐 것 같아요.

mmm307955 replied:
말씀을 듣고 보니 `Money` 클래스가 로또 구입 금액을 의미하는지, 당첨금을 의미하는지 역할이 모호했던 것 같아요. 그래서 로또 구입 금액만을 표현하는 `BuyAmount` 클래스를 새로 만들고, 해당 검증 로직도 그쪽으로 옮기려 합니다.
당첨금은 기존에 분리해둔 `Prize` 클래스를 이용하려고 해요!

seokjin8678 reviewed (changes requested):
안녕하세요 민지님! 로또 미션 2단계 구현하시느라 고생하셨습니다!

지난 미션들에서 배우신 내용 잊지 않고, 꼼꼼하게 테스트 잘 작성해 주셨네요! 👍👍

다만 로또 게임에서 `LottoNumbers`가 중심이 되는 것 같네요.

로또 번호가 중심으로 돌아가는 게 맞지만, 당첨 번호가 `LottoNumbers` 타입인 게 적절한지 고민해 봤으면 좋겠어요.
(LottoNumbers와 LottoTicket의 차이 또한)

해당 사항 반영해 주시고, 재요청 부탁드립니다!

---

> 수익률 결과가 0.35처럼 소수점 이하 두 자리로 출력되는데, 단순히 소수점 아래를 자른 걸까요?
(현재는 %.2f 포맷팅으로 출력하고 있습니다!)

그렇지 않을까요? 출력 포매팅에 관련된 부분은 사소한 곳이니 크게 신경 쓰지 않으셔도 될 것 같네요!

seokjin8678 reviewed (commented):
안녕하세요 민지님! 아직 재요청은 주시지 않았지만, 리뷰에 남긴 답변에 대해 답글 남겨두었습니다!

제출하시기 전 반영해 주시거나, 이미 반영 중이시면 다음 제출 때 반영해 주시면 될 것 같습니다!

mmm307955 committed:
refactor: 2단계 리뷰 사항 반영

- 'Prize' 클래스와 'Rank' enum 클래스를 추가했습니다.
- 'LottoResult.calculateTotalPrize()'의 파라미터 타입을 'Map<Integer, Integer>'에서 'MatchResult'로 변경하고, 이에 따라 'MatchResult' 클래스를 새로 추가했습니다.
- 역할이 모호했던 'Money' 클래스를 제거하고, 'BuyAmount'와 'Prize' 클래스로 역할을 분리했습니다.
- 당첨 번호를 별도로 다루기 위해 'WinningNumbers' 클래스를 추가했습니다.
- 전체 흐름을 담당하는 'LottoController' 클래스를 도입했습니다.

mmm307955 committed:
refactor: 2단계 리뷰 사항 반영

- 'Prize' 클래스와 'Rank' enum 클래스를 추가했습니다.
- 'LottoResult.calculateTotalPrize()'의 파라미터 타입을 'Map<Integer, Integer>'에서 'MatchResult'로 변경하고, 이에 따라 'MatchResult' 클래스를 새로 추가했습니다.
- 역할이 모호했던 'Money' 클래스를 제거하고, 'BuyAmount'와 'Prize' 클래스로 역할을 분리했습니다.
- 당첨 번호를 별도로 다루기 위해 'WinningNumbers' 클래스를 추가했습니다.
- 전체 흐름을 담당하는 'LottoController' 클래스를 도입했습니다.

seokjin8678 commented on src/main/java/view/OutputView.java:
```diff
@@ -17,20 +19,26 @@ public static void printTicketCount(int count) {
 
-    public static void printResult(Map<Integer, Integer> matchResults) {
+    private static String formatTicket(LottoTicket ticket) {
+        return "[" + ticket.getNumbers().stream()
+            .map(String::valueOf)
+            .collect(Collectors.joining(", ")) + "]";
```
`Collectors::joining` 메서드에는 prefix와 suffix를 붙일 수 있답니다.

```suggestion
        return ticket.getNumbers().stream()
            .map(String::valueOf)
            .collect(Collectors.joining(", ", "[", "]"));
```

팁으로 맥북 기준 `CMD` + `P`를 눌러 메서드의 매개변수로 어떤게 들어갈 수 있는지 확인할 수 있어요.

mmm307955 replied:
아하 적용했더니 훨씬 깔끔해졌어요! 매개변수도 앞으로 자주 확인보면 좋을 것 같아요!

seokjin8678 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -7,26 +7,20 @@ private LottoResult() {
-        int totalPrize = 0;
-
-        for (Integer matchCount : matchResults.keySet()) {
-            int prizePerTicket = PRIZE_TABLE.getOrDefault(matchCount, 0);
-            int count = matchResults.get(matchCount);
-            totalPrize += prizePerTicket * count;
+        for (Map.Entry<Rank, Integer> entry : matchResults.getResult().entrySet()) {
+            Rank rank = entry.getKey();
+            int count = entry.getValue();
+            totalPrize = totalPrize.add(rank.getPrize().multiply(count));
         }
 
         return totalPrize;
     }
```
`MatchResult`의 `calculateTotalPrize` 메서드와 차이점이 있나요?

mmm307955 replied:
헉... 두 메서드가 동일한 기능을 하고 있는 중복 코드였네요! 
`MatchResult`를 통해 상금을 계산할 수 있으므로
`LottoResult`의 `calculateTotalPrize()`는 불필요하다고 판단해서 삭제했습니다!
알려주셔서 감사합니다 🙇‍♀️

seokjin8678 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -7,26 +7,20 @@ private LottoResult() {
 
-    public static double calculateRateOfReturn(int totalPrize, int paidMoney) {
-        return (double) totalPrize / paidMoney;
+    public static double calculateRateOfReturn(Prize totalPrize, int paidMoney) {
+        return (double) totalPrize.getAmount() / paidMoney;
     }
```
`Prize` 클래스가 해당 로직을 수행해도 되지 않을까요?

mmm307955 replied:
앗 그렇네요! `Prize` 클래스에 `add`와 `multiply` 메서드가 있는 것 처럼 
상금과 지불 금액으로 수익률을 계산하는 로직이 같이 있어도 자연스러울 것 같아서 `calculateRateOfReturn()` 를 해당 클래스로 옮겼습니다. 

그에 따라 역할이 없어진 `LottoResult` 클래스는 삭제했습니다!

seokjin8678 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -7,26 +7,20 @@ private LottoResult() {
-        6, 2000000000
-    );
+    //로또 당첨 결과: Rank별 당첨 개수
+    public static Prize calculateTotalPrize(MatchResult matchResults) {
```
비즈니스 로직이 `static` 메서드에 구현되고 있네요.

여기 있는 모든 로직은 상태를 가지고 있는 도메인 클래스에 모두 구현할 수 있어 보여요.

이렇게 비즈니스 로직들이 분산되어 있다면 어떤 문제가 발생할 것 같나요?

mmm307955 replied:
우선 비즈니스 로직이 `static` 메서드로 분산되어있으면 객체가 본인 책임에 집중하지 못해 응집도가 떨어지고 책임이 불명확해질 수 있을 것 같아요. 또한 로직이 흩어져있으면 추후 유지보수 시 수정해야할 곳이 어딘지도 직관적으로 파악하기도 힘들 것 같아요. 현재 해당 클래스는 삭제했습니다!

seokjin8678 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -12,12 +12,12 @@ public LottoTickets(List<LottoTicket> tickets) {
-    public Map<Integer, Integer> countMatchResults(LottoNumbers winningNumbers) {
-        Map<Integer, Integer> results = new HashMap<>();
+    public MatchResult countMatchResults(WinningNumbers winningNumbers) {
+        Map<Rank, Integer> results = new EnumMap<>(Rank.class);
```
`EnumMap`을 사용해 주셨군요! 👍👍

HashMap과 어떤 차이가 있을까요?

다른 `Map` 구현체에 대해서도 알아보시면 좋을 것 같아요.

mmm307955 replied:
넵! 인텔리제이에서 `Map` 인터페이스쪽 살짝 봤는데 `TreeMap`도 있고 `SortedMap`이라는 것도 있네요! 나중에 더 깊게 찾아봐야겠어요!

아래는 그냥 중얼거림입니다! 

우선 `EnumMap`이 `HashMap`보다 빠르다고는 대략만 알고 있었는데, 왜 빠른지는 몰라서 이번에 찾아보았어요. 내부적으로 `enum`의 `ordinal()`로 인덱스로 써서 순번을 갖기 때문에 해싱도 안 하고 해싱충돌을 처리할 필요도 없어서, 그리고 리사이징을 할 필요가 없어서 빠르다는 것을 알게 되었어요. `ordinal()`이 있는 건 알았는데, 이게 인덱스를 저장할 때에도 쓰이는지는 건지는 몰랐어요!

HashMap은 키의 해시값을 인덱스로 사용하고, 해싱 충돌이 발생할 수도 있다는 걸 알게 되었어요. 그리고 자바에서 이걸 해결하기 위해서 체이닝이라는 걸로 해결하는것도 알게 되었어요.

뭔가 궁금해지는 것도 많고, 딴길로 많이 샌 것 같지만 찾아보면서 재밌었어요..!

seokjin8678 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -12,12 +12,12 @@ public LottoTickets(List<LottoTicket> tickets) {
+    public MatchResult countMatchResults(WinningNumbers winningNumbers) {
+        Map<Rank, Integer> results = new EnumMap<>(Rank.class);
         for (LottoTicket ticket : tickets) {
-            int match = ticket.countMatch(winningNumbers);
-            results.put(match, results.getOrDefault(match, 0) + 1);
+            Rank rank = ticket.countMatch(winningNumbers);
+            results.put(rank, results.getOrDefault(rank, 0) + 1);
         }
-        return results;
+        return new MatchResult(results);
```
억지로 구현한 점은 있지만.. Stream을 사용하면 다음과 같이 표현할 수 있어요.

```java
Map<Rank, Integer> result = tickets.stream()
    .map(ticket -> ticket.countMatch(winningNumbers))
    .collect(groupingBy(Function.identity(), () -> new EnumMap<>(Rank.class), summingInt(e -> 1)));
return new MatchResult(result);
```

mmm307955 replied:
와 처음 보는 것들이 많은데... 이렇게 쓰는군요... 덕분에 배워갑니다 👍

seokjin8678 commented on src/test/java/domain/LottoResultTest.java:
```diff
@@ -12,25 +12,25 @@ class LottoResultTest {
-            4, 1, // 번호 4개 맞은 게 1장 => 50000 * 1 = 50000
-            5, 0, // 번호 5개 맞은 게 0장 => 1500000 * 0 = 0
-            6, 0 // 번호 6개 맞은 게 0장 => 2000000000 * 0 = 0
-        );
+        MatchResult matchResult = new MatchResult(Map.of(
+            Rank.FOURTH, 2,
+            Rank.THIRD, 0,
+            Rank.SECOND, 0,
+            Rank.FIRST, 0
+        ));
 
         //When
-        int totalPrize = LottoResult.calculateTotalPrize(matchResult);
+        Prize totalPrize = LottoResult.calculateTotalPrize(matchResult);
 
         //Then
-        assertThat(totalPrize).isEqualTo(60000);
+        assertThat(totalPrize).isEqualTo(10000);
     }
```
테스트 코드에 무엇이 잘못되었나요?

mmm307955 replied:
> 이 또한 자동화할 수 있지 않을까요?
말씀해주셔서 test 폴더에서 `Run 'All Tests'`를 통해 테스트해보았더니 3개나 실패하는 것을 확인했습니다...!
앞으로 이렇게 제출 전 체크하면 편리하게 테스트할 수 있을 것 같습니다!

```
expected: 10000
 but was: domain.Prize@2710
Expected :10000
Actual   :domain.Prize@2710
```
Prize는 객체인데 기본타입으로 직접 비교하여 테스트가 실패했습니다. 

`assertThat(totalPrize).isEqualTo(Prize.from(10000));`
로 수정하였습니다!

seokjin8678 commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,30 @@
+        return prize;
+    }
+
+    public static Rank from(int matchCount) {
+        return Arrays.stream(values())
+            .filter(result -> result.matchCount == matchCount)
+            .findFirst()
+            .orElse(NONE);
+    }
```
특정 값에 대한 `Enum`을 구하기 위해 구현하신 정적 팩터리 메서드네요!

이러한 방식은 거의 패턴처럼 사용되며, 쉽게 구현할 수 있어요.

하지만 이러한 방식은 문제점을 하나 가지고 있어요.

```java
Rank[] first = Rank.values();
Rank[] second = Rank.values();

assertThat(first == second).isTrue();
```
해당 테스트의 결과는 어떻게 될까요?

왜 그런지 고민을 해보시고, 이유를 알고 싶다면 해당 [링크](https://stackoverflow.com/questions/32799282/does-the-enumvalues-allocate-memory-on-each-call)를 참고하시면 좋을 것 같아요.

mmm307955 replied:
음 처음에는 단순히 두 배열은 다른 주소값을 가르키기 때문에 테스트가 실패할 것 같다고 생각했습니다. 그런데 불변성 보장을 위해 `values()`가 호출될 때마다 새로운 배열을 반환하기 때문에 테스트가 실패하는 것이었어요. 따라서 참조 동일성을 보장하기 위해 `values()` 호출 결과를 `Collections.unmodifiableList()`로 감싼 `VALUES` 리스트를 만들어 `getValues()`로 반환하도록 수정했습니다!

seokjin8678 commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Arrays;
+
+public enum Rank {
+    NONE(0, Prize.from(0L)),
+    FOURTH(3, Prize.from(5_000L)),
+    THIRD(4, Prize.from(50_000L)),
+    SECOND(5, Prize.from(1_500_000L)),
+    FIRST(6, Prize.from(2_000_000_000L));
```
`_`로 잘 나눠주셔서 가독성이 좋군요 👍👍

seokjin8678 commented on src/main/java/domain/MatchResult.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.Map;
+
+public class MatchResult {
+    private final Map<Rank, Integer> matchCountByRank;
+
+    public MatchResult(Map<Rank, Integer> matchCountByRank) {
+        this.matchCountByRank = matchCountByRank;
+    }
+
+    public int getCount(Rank rank) {
+        return matchCountByRank.getOrDefault(rank, 0);
+    }
+
+    public Prize calculateTotalPrize() {
+        Prize total = Prize.from(0L);
+        for (Rank rank : Rank.values()) {
+            Prize prize = rank.getPrize().multiply(getCount(rank));
+            total = total.add(prize);
+        }
+        return total;
+    }
+
+    public Map<Rank, Integer> getResult() {
+        return Map.copyOf(matchCountByRank);
+    }
+}
```
기존 `Map<Integer, Integer>`를 사용한 방식보다 훨씬 더 이해하기 쉬워졌네요!

mmm307955 replied:
휴 다행이에요!!

seokjin8678 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,52 @@
+
+public class LottoController {
+    public void run() {
+        //로또들을 구매하고 티켓을 생성한다
+        BuyAmount buyAmount = new BuyAmount(InputView.inputMoney());
+        List<LottoTicket> lottoTickets = buyLottoTickets(buyAmount);
+
+        //구매한 로또들을 출력한다
+        OutputView.printTickets(lottoTickets);
+
+        //저번 주 당첨번호를 입력한다
+        WinningNumbers winningNumbers = new WinningNumbers(InputView.inputWinningNumbers());
+
+        //등수에 따라 몇 개가 맞았는지 출력한다
+        MatchResult matchResult = calculateMatchResult(lottoTickets, winningNumbers);
+        OutputView.printResult(matchResult);
+
+        //수익률을 계산하고 출력한다
+        printProfit(matchResult, buyAmount);
```
코드의 길이는 10줄을 넘겼지만, 주석을 지우면 10줄이 되지 않을 것 같네요.

하지만 요구사항을 맞춘다고 억지로 10줄로 만든 것보다, 이처럼 주석을 남겨서 10줄을 넘긴 게 더 가독성이 높은 것 같네요!

mmm307955 replied:
감사합니다!

seokjin8678 commented on src/main/java/domain/MatchResult.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public Prize calculateTotalPrize() {
+        Prize total = Prize.from(0L);
+        for (Rank rank : Rank.values()) {
+            Prize prize = rank.getPrize().multiply(getCount(rank));
+            total = total.add(prize);
+        }
+        return total;
```
마찬가지로, 억지로 구현한 느낌은 있지만, Stream을 사용한다면 다음과 같이 표현할 수 있어요.
```java
return matchCountByRank.keySet().stream()
    .map(rank -> rank.getPrize().multiply(getCount(rank)))
    .reduce(Prize::add)
    .orElseGet(() -> Prize.from(0));
```

mmm307955 replied:
`.reduce(Prize::add)` 이렇게 쓸 수도 있군요..!

seokjin8678 commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,29 @@
+        return winningNumbers;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof WinningNumbers that)) {
+            return false;
+        }
+        return Objects.equals(winningNumbers, that.winningNumbers);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(winningNumbers);
+    }
```
정말 동등성을 보장할까요?

```java
WinningNumbers first = new WinningNumbers(List.of(1, 2, 3, 4, 5, 6));
WinningNumbers second = new WinningNumbers(List.of(1, 2, 3, 4, 5, 6));

assertThat(first).isEqualTo(second);
```

mmm307955 replied:
앗!! 몰랐습니다... LottoNumbers에도 오버라이딩하여 해결했습니다!

seokjin8678 reviewed (approved):
안녕하세요 민지님! 리뷰 반영하신 것 확인했습니다!

전체적으로 객체를 잘 만들어 주셨고, 테스트 코드 또한 꼼꼼하게 잘 작성해 주셨습니다. 👍

다만 일부 테스트 코드가 실패하는 것 같아요. 😂

코드 스타일과 마찬가지로, 제출하기 전 모든 테스트를 실행해 보는 습관을 들여보는 게 좋을 것 같아요.
> 이 또한 자동화할 수 있지 않을까요?

요구사항은 전부 만족하셨으니, 5단계까지 구현하시면 될 것 같습니다!
구현하시고 요청 부탁드립니다!
